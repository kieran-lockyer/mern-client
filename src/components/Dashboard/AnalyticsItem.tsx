import React from "react";
import { Icon, Colors } from "@blueprintjs/core";
import styled from "styled-components";

export default ({ tag, count }) => {
  return (
    <CardItem>
      <CardLeft>
        <Icon color={Colors.LIGHT_GRAY1} icon="tag" iconSize={25} />
      </CardLeft>
      <CardDetails>
        <CardTitle>{tag}</CardTitle>
        <CardCount>{count}</CardCount>
      </CardDetails>
    </CardItem>
  );
};

const CardItem = styled.div`
  display: flex;
  align-items: center;
  min-height: 5rem;
  padding: 0 1rem;

  &:last-child {
    border-bottom-left-radius: inherit;
    border-bottom-right-radius: inherit;
  }

  &:hover {
    background: #f9f9f9;
  }
`;

const CardLeft = styled.div`
  margin-right: 1rem;
`;

const CardDetails = styled.div`
  margin-right: auto;
`;

const CardTitle = styled.h3`
  font-weight: bold;
  margin: 0;
  font-size: 15px;
`;

const CardCount = styled.p`
  font-size: 0.875rem;
  color: #ccc;
`;
