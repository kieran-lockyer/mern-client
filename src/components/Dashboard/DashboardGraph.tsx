import React, { Component } from "react";
import Chart from "react-apexcharts";
import { ButtonGroup, Button } from "@blueprintjs/core";
import { connect } from "react-redux";
import * as actions from "../../actions";
import styled from "styled-components";
import { Cube } from "react-preloaders";

class DashboardGraph extends Component<any, any> {
  render() {
    if (this.props.selection) {
      return (
        <div>
          <ButtonGroup fill>
            <Button
              onClick={() => this.props.fetchGraphData(7, "one_week")}
              id="one_week"
              className={
                this.props.selection === "one_week"
                  ? "date-button-group date-button-group-active"
                  : "date-button-group"
              }
            >
              1W
            </Button>
            <Button
              onClick={() => this.props.fetchGraphData(30, "one_month")}
              id="one_month"
              className={
                this.props.selection === "one_month"
                  ? "date-button-group date-button-group-active"
                  : "date-button-group"
              }
            >
              1M
            </Button>
            <Button
              onClick={() => this.props.fetchGraphData(90, "three_months")}
              id="three_months"
              className={
                this.props.selection === "three_months"
                  ? "date-button-group date-button-group-active"
                  : "date-button-group"
              }
            >
              3M
            </Button>
            <Button
              onClick={() => this.props.fetchGraphData(180, "six_months")}
              id="six_months"
              className={
                this.props.selection === "six_months"
                  ? "date-button-group date-button-group-active"
                  : "date-button-group"
              }
            >
              6M
            </Button>
          </ButtonGroup>
          <Chart options={this.props.options} series={this.props.series} />
        </div>
      );
    } else {
      return (
        <Container>
          <Cube color={"#48c0b9"} bgColor={"transparent"} />
        </Container>
      );
    }
  }
}

const Container = styled.div`
  flex: 1 1 100%;
  padding: 30px 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const mapStateToProps = state => ({
  options: state.graph.options,
  series: state.graph.series,
  selection: state.graph.selection
});

export default connect(
  mapStateToProps,
  actions
)(DashboardGraph);
