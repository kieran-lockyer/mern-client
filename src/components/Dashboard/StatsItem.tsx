import React from "react";
import styled from "styled-components";
import { Icon } from "@blueprintjs/core";

const StatsItem = props => {
  const { avgTags, avgPhotos } = props;
  if (avgTags || avgPhotos) {
    return (
      <AvgStats>
        <StatsIcon>
          <Icon icon={props.icon} iconSize={100} />
        </StatsIcon>
        <StatsDetails>
          {avgTags && <h2>Avg. Daily Tags</h2>}
          {avgPhotos && <h2>Avg. Daily Photos</h2>}
          <span>{avgTags || avgPhotos}</span>
        </StatsDetails>
      </AvgStats>
    );
  } else {
    return (
      <Stats>
        <Icon icon={props.icon} iconSize={75} />
        <span>{props.totalPhotos || props.totalTags}</span>
      </Stats>
    );
  }
};

const Stats = styled.div`
  padding: 2rem;
  background: #fff;
  flex-grow: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  border-radius: 10px;
  flex-direction: column;
  color: #fff;
  background: linear-gradient(to right bottom, #223f67, #192b44);
  & h2 {
    font-size: 30px;
  }

  & > span {
    padding: 10px;
    width: 100%;
    font-weight: bold;
    text-transform: uppercase;
    text-align: center;
    border-radius: inherit;
    font-size: 14px;
    margin-top: 10px;
    justify-content: center;
    display: flex;
    color: #51c7c1;
    &:last-child {
      background: #0000004f;
    }
  }

  &:not(:last-child) {
    margin-bottom: 0.9rem;
  }
`;

const AvgStats = styled.div`
  flex: 1;
  display: flex;
  justify-content: space-around;
  align-items: center;
  background: linear-gradient(to bottom, #223f67, #192b44);
  border-radius: 15px;
  color: white;
  padding: 1rem;
  position: relative;

  &:nth-child(1) {
    margin-right: 10px;

    @media only screen and (max-width: 800px) {
      margin-right: 0;
      margin-bottom: 1rem;
    }
  }

  h2 {
    margin: 0;
  }
`;

const StatsIcon = styled.div`
  opacity: 0.2;
  position: absolute;
  left: -10px;
  top: -10px;
`;

const StatsDetails = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin: 1rem 0;
`;

export default StatsItem;
