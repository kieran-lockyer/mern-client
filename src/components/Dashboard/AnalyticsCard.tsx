import React, { Component } from "react";
import AnalyticsItem from "./AnalyticsItem";
import styled from "styled-components";
import { connect } from "react-redux";

class AnalyticsCard extends Component<any, any> {
  renderPopTags = () => {
    return this.props.popular.map(item => {
      return <AnalyticsItem tag={item._id} count={item.count} key={item._id} />;
    });
  };

  renderTrendingTags = () => {
    return this.props.trending.map(item => {
      return <AnalyticsItem tag={item._id} count={item.count} key={item._id} />;
    });
  };

  renderCard = () => {
    return this.props.title === "Most Popular Tags"
      ? this.renderPopTags()
      : this.renderTrendingTags();
  };

  render() {
    return (
      <Card>
        <CardHeading>
          <h3>{this.props.title}</h3>
        </CardHeading>
        {this.renderCard()}
      </Card>
    );
  }
}

const Card = styled.div`
  background: #fff;
  flex: 1 1 48%;
  margin: 10px 10px 30px 0;
  border-radius: 15px;
  align-self: start;
  &:nth-child(2n),
  &:last-child {
    margin-right: 0;
  }

  @media only screen and (max-width: 1000px) {
    flex-basis: 100%;
    margin: 0 0 1rem 0;
  }
`;

const CardHeading = styled.div`
  color: #fff;
  background: linear-gradient(to right bottom, #223f67, #192b44);
  border-top-left-radius: inherit;
  border-top-right-radius: inherit;
  border-bottom: 1px solid #eee;
  padding: 1rem 2rem;
  display: flex;
  align-items: center;

  h3 {
    margin: 0;
  }
`;

const mapStateToProps = state => ({
  popular: state.stats.popTags,
  trending: state.stats.trendingTags
});

export default connect(mapStateToProps)(AnalyticsCard);
