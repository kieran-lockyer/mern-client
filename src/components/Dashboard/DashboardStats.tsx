import React from "react";
import StatsItem from "./StatsItem";
import { connect } from "react-redux";
import styled from "styled-components";

const DashboardStats = props => {
  return (
    <Stats>
      <StatsItem icon="media" totalPhotos={props.totalPhotos} />
      <StatsItem icon="tag" totalTags={props.totalTags} />
    </Stats>
  );
};
const Stats = styled.div`
  flex-grow: 1;
  margin-left: 2rem;
  justify-content: space-between;
  display: flex;
  flex-direction: column;
  @media only screen and (max-width: 800px) {
    margin-left: 0;
  }
`;

const mapStateToProps = state => ({
  totalPhotos: state.photos.photoData.totalDocs,
  totalTags: state.tags.tagData.totalDocs
});

export default connect(mapStateToProps)(DashboardStats);
