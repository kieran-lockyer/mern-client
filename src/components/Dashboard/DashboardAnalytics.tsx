import React from "react";
import AnalyticsCard from "./AnalyticsCard";
import StatsItem from "./StatsItem";
import styled from "styled-components";
import { connect } from "react-redux";

const DashboardAnalytics = props => {
  return (
    <>
      <AvgStats>
        <StatsItem icon="media" avgPhotos={props.avgPhotos} />
        <StatsItem icon="tag" avgTags={props.avgTags} />
      </AvgStats>
      <Container>
        <AnalyticsCard title="Popular Tags" />
        <AnalyticsCard title="Trending Tags" />
      </Container>
    </>
  );
};

const Container = styled.section`
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
`;

const AvgStats = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
  margin-bottom: 10px;
  @media only screen and (max-width: 800px) {
    flex-direction: column;
  }
`;

const mapStateToProps = state => ({
  avgTags: state.stats.avgTags,
  avgPhotos: state.stats.avgPhoto
});

export default connect(mapStateToProps)(DashboardAnalytics);
