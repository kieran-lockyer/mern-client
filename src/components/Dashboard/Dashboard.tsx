import React, { Component } from "react";
import DashboardAnalytics from "./DashboardAnalytics";
import DashboardGraph from "./DashboardGraph";
import DashboardStats from "./DashboardStats";
import styled from "styled-components";
import { connect } from "react-redux";
import * as actions from "../../actions";
import { Cube } from "react-preloaders";

class Dashboard extends Component<any, any> {
  componentDidMount() {
    try {
      this.props.fetchPhotos(1, 30, "dateAdded", "desc", "");
      this.props.fetchTags(1, 30, "dateAdded", "desc", "");
      this.props.fetchStats();
      this.props.fetchGraphData(30, "one_month");
    } catch {
      err => console.log(err);
    }
  }

  render() {
    const { isLoading } = this.props;

    if (!isLoading) {
      return (
        <Container>
          <Wrapper>
            <Header>
              <Graph>
                <DashboardGraph />
              </Graph>
              <DashboardStats />
            </Header>
            <DashboardAnalytics />
          </Wrapper>
        </Container>
      );
    } else {
      return (
        <Container>
          <Cube color={"#48c0b9"} bgColor={"transparent"} />
        </Container>
      );
    }
  }
}

// Styled Components
const Container = styled.div`
  flex: 1 1 100%;
  padding: 30px 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Wrapper = styled.div`
  width: 85%;
  padding: 2rem;
  max-width: 1300px;

  @media only screen and (max-width: 800px) {
    width: 100%;
    padding: 1rem;
  }
`;

const Header = styled.section`
  border-radius: 15px;
  margin-bottom: 3rem;
  display: flex;
  @media only screen and (max-width: 800px) {
    flex-direction: column;
  }
`;

const Graph = styled.div`
  background: #18263d;
  border-radius: 1rem;
  box-shadow: 0 0 14px 12px #e8e8e8;
  padding: 20px;
  box-sizing: border-box;
  width: 75%;
  margin-right: auto;
  @media only screen and (max-width: 800px) {
    width: 100%;
    margin-bottom: 2rem;
  }
`;

const mapStateToProps = state => ({
  isLoading: state.stats.isLoading
});

export default connect(
  mapStateToProps,
  actions
)(Dashboard);
