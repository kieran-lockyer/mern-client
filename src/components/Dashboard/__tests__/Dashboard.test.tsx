import React from 'react';
import { shallow } from 'enzyme';
import Dashboard from '../Dashboard';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

describe('<Dashboard />', () => {
  describe('render()', () => {
    test('renders the Dashboard component', () => {
      const component = shallow(<Dashboard />);

      expect(toJson(component)).toMatchSnapshot();
    });
  });
});