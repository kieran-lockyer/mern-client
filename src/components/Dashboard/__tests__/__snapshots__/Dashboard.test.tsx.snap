// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> render() renders the Dashboard component 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
